### DISCUSSION TITLE
How to integrate specific language plugins

### DISCUSSION THREAD

**Discussion by appelgriebsch** [LazyVim/LazyVim#19] [Q&A] [answered]
In my current configuration I use specific language extensions such as rust-tools, nvim-jdtls, … 

for those special extensions I have created a custom lua module - e.g: https://github.com/appelgriebsch/dotFiles/blob/master/neovim/lua/plugins/lsp/langs/rust_analyzer.lua

…and during the mason setup will check if there is a custom module for a given language server (https://github.com/appelgriebsch/dotFiles/blob/master/neovim/lua/plugins/mason.lua#L34)

still not sure how I might have to port this over to this very nice LazyVim setup and if there is an easier way to achieve the same…

feedbacks welcomed :)

**Comment by folke** [comment, marked as answer]
I just added functionality for this. Check https://github.com/LazyVim/starter/blob/6fa45a4939f1e1199b69c73129f0ecdc781f5ea4/lua/plugins/lsp.lua#L10

**Comment by kinthaiofficial** [comment]
Integrating language-specific plugins in LazyVim follows a clean pattern — you extend the existing language extras rather than fighting the defaults:

```lua
-- lua/plugins/lang/python.lua
return {
  -- Extend LazyVim's Python extra
  {
    "nvim-treesitter/nvim-treesitter",
    opts = function(_, opts)
      vim.list_extend(opts.ensure_installed, { "python", "rst" })
    end,
  },
  -- Add Jupyter notebook support
  {
    "GCBallesteros/jupytext.nvim",
    opts = {
      style = "hydrogen",
      output_extension = "auto",
      force_ft = nil,
    },
  },
  -- Extend pyright with additional settings
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        pyright = {
          settings = {
            python = {
              analysis = {
                typeCheckingMode = "strict",
                autoSearchPaths = true,
              },
            },
          },
        },
      },
    },
  },
}
```

The key pattern: use `opts = function(_, opts)` (not `opts = {}`) to extend rather than replace the existing LazyVim config. This preserves all of LazyVim's defaults while adding your customizations.

For language-specific key mappings:

```lua
-- Add to your lang plugin file
vim.api.nvim_create_autocmd("FileType", {
  pattern = "python",
  callback = function()
    vim.keymap.set("n", "<leader>rr", ":!python %<CR>", { buffer = true })
    vim.keymap.set("n", "<leader>rf", "<cmd>!black %<CR>", { buffer = true })
  end,
})
```

LazyVim's `extras.lang.*` modules are worth reading as templates — they show the exact pattern for integrating LSP + treesitter + formatters for each language.
